File: platform/src/arch/aarch64/cpu_local.c
#include "cpu_local.h"

static inline void msr_tpidr_el1_write(uintptr_t ptr) {
    __asm__ volatile("msr tpidr_el1, %0" : : "r"(ptr));
}

static inline uintptr_t msr_tpidr_el1_read(void) {
    uintptr_t ptr;
    __asm__ volatile("mrs %0, tpidr_el1" : "=r"(ptr));
    return ptr;
}

static struct cpu_local_storage g_cpu_local = {
    .task_id = 0,
};

void cpu_local_init(void) {
    msr_tpidr_el1_write((uintptr_t) &g_cpu_local);
}

struct cpu_local_storage *cpu_local_get(void) {
    const uintptr_t ptr = msr_tpidr_el1_read();
    return (struct cpu_local_storage *) ptr;
}

File: kernel/old_kernel/hal/src/arch/aarch64/cpu_local.rs
use core::arch::asm;
use core::ptr::{NonNull, null_mut};

unsafe fn msr_tpidr_el1_write(ptr: NonNull<CpuLocalStorage>) {
    let ptr = ptr.as_ptr();
    asm!("msr tpidr_el1, {}", in(reg) ptr);
}

unsafe fn msr_tpidr_el1_read() -> Option<NonNull<CpuLocalStorage>> {
    let ptr: *mut CpuLocalStorage;
    asm!("mrs {}, tpidr_el1", out(reg) ptr);
    NonNull::new(ptr)
}

#[repr(C)]
pub struct CpuLocalStorage {
    pub(crate) task_id: u64,
}

static mut CPU_LOCAL: CpuLocalStorage = CpuLocalStorage { task_id: 0 };

pub unsafe fn init() {
    unsafe {
        let ptr = &raw mut CPU_LOCAL;
        let ptr = NonNull::new(ptr).unwrap();
        msr_tpidr_el1_write(ptr);
    }
}

pub unsafe fn get() -> Option<NonNull<CpuLocalStorage>> {
    unsafe { msr_tpidr_el1_read() }
}
